certora/specs/v1.5/Guards.spec

```diff
@@ -0,0 +1,212 @@
+
+    function _.checkTransaction(
+        address to,
+        uint256 value,
+        bytes data,
+        Enum.Operation operation,
+        uint256 safeTxGas,
+        uint256 baseGas,
+        uint256 gasPrice,
+        address gasToken,
+        address refundReceiver,
+        bytes signatures,
+        address msgSender
+    ) external => DISPATCHER(true) ;
+
+    function _.checkAfterExecution(bytes32 hash, bool success) external 
+        => DISPATCHER(true);
+
+    function Executor.execute(
+        address to,
+        uint256 value,
+        bytes memory data,
+        Enum.Operation operation,
+        uint256 txGas
+    ) internal returns (bool) => NONDET;
+
+    function SecuredTokenTransfer.transferToken(address token, address receiver, uint256 amount) internal returns (bool) => NONDET ;
+    function Safe.handlePayment(
+        uint256 gasUsed,
+        uint256 baseGas,
+        uint256 gasPrice,
+        address gasToken,
+        address refundReceiver
+    ) internal returns (uint256) => NONDET ;
+
+}
+
+// ---- Functions and ghosts ---------------------------------------------------
+
+
+// ---- Invariants -------------------------------------------------------------
+
+
+// ---- Rules ------------------------------------------------------------------
+
+/// @dev the only method that can change the guard is setGuard
+/// @status Done: https://prover.certora.com/output/39601/a05e24787c68404d877ae4acce693734?anonymousKey=02030d2ca97a19d0d7a70deb5a91dc4b75bca89d
+rule guardAddressChange(method f) filtered {
+    f -> f.selector != sig:simulateAndRevert(address,bytes).selector &&
+         f.selector != sig:getStorageAt(uint256,uint256).selector
+} {
+    address guardBefore = getSafeGuard();
+
+    calldataarg args; env e;
+    f(e, args);
+
+    address guardAfter = getSafeGuard();
+
+    assert guardBefore != guardAfter =>
+        f.selector == sig:setGuard(address).selector;
+}
+
+/// @dev the only method that can change the module guard is setModuleGuard
+/// @status Done: https://prover.certora.com/output/39601/a05e24787c68404d877ae4acce693734?anonymousKey=02030d2ca97a19d0d7a70deb5a91dc4b75bca89d
+
+rule moduleGuardAddressChange(method f) filtered {
+    f -> f.selector != sig:simulateAndRevert(address,bytes).selector &&
+         f.selector != sig:getStorageAt(uint256,uint256).selector
+} {
+    address guardBefore = getModuleGuardExternal();
+    
+    calldataarg args; env e;
+    f(e,args);
+    
+    address guardAfter = getModuleGuardExternal();
+
+    assert guardBefore != guardAfter => 
+        f.selector == sig:setModuleGuard(address).selector;
+}
+
+/// @dev set-get correspondence for (regular) guard
+/// @status Done: https://prover.certora.com/output/39601/a05e24787c68404d877ae4acce693734?anonymousKey=02030d2ca97a19d0d7a70deb5a91dc4b75bca89d
+rule setGetCorrespondenceGuard(address guard) {
+    env e;
+    setGuard(e,guard);
+    address gotGuard = getSafeGuard();
+    assert guard == gotGuard;
+}
```

Comment:
I wonder if we should also check that the guard must implement the guard interface.
`ITransactionGuard(guard).supportsInterface(type(ITransactionGuard).interfaceId))`

Not sure if it's possible to call an interface method from CVL though.  Also that only holds if supportsInterface method is pure (always returns the same value), which theoretically doesn't have to be the case.